segmentation/utils/segmentation_utils.py

```diff
@@ -287,59 +367,87 @@ def compute_marker_counts(input_images, segmentation_masks, nuclear_counts=False
-        cell_counts = signal_extraction_utils.default_extraction(cell_coords, input_images)
-        xr_counts.loc['whole_cell', cell.label, xr_counts.features[1]:] = cell_counts
-        xr_counts.loc['whole_cell', cell.label, xr_counts.features[0]] = cell.area
+        cell_counts = signal_extraction.default_extraction(cell_coords, input_images)
+
+        # get morphology metrics
+        current_cell_props = cell_props.loc[cell_props['label'] == cell_id, object_properties[:-1]]
+
+        # combine marker counts and morphology metrics together
+        cell_features = np.concatenate((cell_counts, current_cell_props), axis=None)
+
+        # add counts of each marker to appropriate column
+        marker_counts.loc['whole_cell', cell_id, marker_counts.features[1]:] = cell_features
+
+        # add cell size to first column
+        marker_counts.loc['whole_cell', cell_id, marker_counts.features[0]] = cell_coords.shape[0]
 
         if nuclear_counts:
-            # only include cell_coords that overlap with a nuclear label
-            nuc_coords = []
-            nuc_mask = segmentation_masks.loc[:, :, 'nuclear']
-            for idx in range(cell_coords.shape[1]):
-                if nuc_mask[cell_coords[0, idx], cell_coords[1, idx]] > 0:
-                    nuc_coords.append(cell_coords[:, idx])
+            # get id of corresponding nucleus
+            nuc_id = find_nuclear_mask_id(nuc_segmentation_mask=nuc_mask, cell_coords=cell_coords)
+
+            if nuc_id is None:
+                # no nucleus found within this cell
+                pass
+            else:
+                # get coordinates of corresponding nucleus
+                nuc_coords = nuc_props.loc[nuc_props['label'] == nuc_id, 'coords'].values[0]
+
+                # extract nuclear signal
+                nuc_counts = signal_extraction.default_extraction(nuc_coords, input_images)
+
+                # get morphology metrics
+                current_nuc_props = nuc_props.loc[
+                    nuc_props['label'] == nuc_id, object_properties[:-1]]
 
-            # extract data from nuclear coords
-            nuc_coords = np.stack(nuc_coords, axis=-1)
-            nuc_counts = signal_extraction_utils.default_extraction(nuc_coords, input_images)
-            xr_counts.loc['nuclear', cell.label, xr_counts.features[1]:] = nuc_counts
-            xr_counts.loc['nuclear', cell.label, xr_counts.features[0]] = nuc_coords.shape[1]
+                # combine marker counts and morphology metrics together
+                nuc_features = np.concatenate((nuc_counts, current_nuc_props), axis=None)
 
-    return xr_counts
+                # add counts of each marker to appropriate column
+                marker_counts.loc['nuclear', nuc_id, marker_counts.features[1]:] = nuc_features
+
+                # add cell size to first column
+                marker_counts.loc['nuclear', nuc_id, marker_counts.features[0]] = \
+                    nuc_coords.shape[0]
+
```

Comment:
A lot of the nuclear counting logic is the shared with the whole cell.  I think a lot of this could be moved into a loop over the `segmentation_masks.compartments` (currently from 'whole_cell' to 'nuclear'), which opens the door to easily adding more types of mask counting.  

The only hiccup is associating each type of label mask (nuclear, etc) to each cell, but that functionality is already covered by `find_nuclear_mask_id`.  So, this could be a loop so that for each `c in compartments` apply `find_nuclear_mask_id` if `c is not 'whole_cell'` and then perform the rest of the shared logic, storing the result in `marker_counts.loc[c, nuc_id, markercounts.features[1]:]`

Reply:
also, sorry i only just now remembered to use the 'start a review button' and not single comment.

Reply:
Right, the issue is that find_nuclear_mask_id requires the cell coords for that cell to calculate the right nuclear ID. 

You're totally right that we'll be incorporating additional compartments. Gonna punt on this for now, whichever one of you gets to this first can take a stab at it.